framework/src/interfaces/TaggingInterface.C

```diff
@@ -157,26 +158,58 @@ TaggingInterface::useMatrixTag(TagID tag_id)
 }
 
 void
-TaggingInterface::prepareVectorTag(Assembly & assembly, unsigned int ivar)
+TaggingInterface::prepareVectorTag(Assembly & assembly, const unsigned int ivar)
+{
+  prepareVectorTagInternal(assembly, ivar, _vector_tags, _abs_vector_tags, {}, {});
+}
+
+void
+TaggingInterface::prepareVectorTag(Assembly & assembly,
+                                   const unsigned int ivar,
+                                   const ReferenceResidualProblem * const ref_problem,
+                                   const bool prepare_non_ref_tags)
 {
-  _re_blocks.resize(_vector_tags.size());
   mooseAssert(_vector_tags.size() >= 1, "we need at least one active tag");
-  auto vector_tag = _vector_tags.begin();
-  for (MooseIndex(_vector_tags) i = 0; i < _vector_tags.size(); i++, ++vector_tag)
-  {
-    const VectorTag & tag = _subproblem.getVectorTag(*vector_tag);
-    _re_blocks[i] = &assembly.residualBlock(ivar, tag._type_id);
-  }
 
-  _local_re.resize(_re_blocks[0]->size());
+  const auto ref_tags =
+      ref_problem ? std::set<TagID>({ref_problem->referenceVectorTagID({})}) : std::set<TagID>({});
+  const auto abs_ref_tags = _abs_vector_tags.empty() ? std::set<TagID>({}) : ref_tags;
 
-  _absre_blocks.resize(_abs_vector_tags.size());
-  vector_tag = _abs_vector_tags.begin();
-  for (MooseIndex(_abs_vector_tags) i = 0; i < _abs_vector_tags.size(); i++, ++vector_tag)
+  if (prepare_non_ref_tags)
+    prepareVectorTagInternal(
+        assembly, ivar, _vector_tags, _abs_vector_tags, ref_tags, abs_ref_tags);
+  else
+    prepareVectorTagInternal(assembly, ivar, ref_tags, abs_ref_tags, {}, {});
+}
+
+void
+TaggingInterface::prepareVectorTagInternal(
+    Assembly & assembly,
+    const unsigned int ivar,
+    const std::set<TagID> & vector_tags,
+    const std::set<TagID> & absolute_value_vector_tags,
+    const std::set<TagID> & vector_tags_to_skip,
+    const std::set<TagID> & absolute_value_vector_tags_to_skip)
+{
+  auto prepare =
+      [this, ivar, &assembly](auto & re_blocks, const auto & tags, const auto & tags_to_skip)
   {
-    const VectorTag & tag = _subproblem.getVectorTag(*vector_tag);
-    _absre_blocks[i] = &assembly.residualBlock(ivar, tag._type_id);
-  }
+    re_blocks.clear();
+    re_blocks.reserve(tags.size());
+    for (const auto tag_id : tags)
+    {
+      if (tags_to_skip.count(tag_id))
+        continue;
```

Comment:
Did this exist before? The skipping? I think not. Seems like it would be better to filter before hand than putting this in a loop, but... I'll just let that one simmer for you. I'm okay either way.

Reply:
Changed